webauthn/src/SafeWebAuthnPlugin.sol

```diff
@@ -0,0 +1,91 @@
+// SPDX-License-Identifier: LGPL-3.0-only
+pragma solidity >=0.7.0 <0.9.0;
+pragma abicoder v2;
+
+import {BaseAccount} from "account-abstraction/contracts/core/BaseAccount.sol";
+import {IEntryPoint, UserOperation} from "account-abstraction/contracts/interfaces/IEntryPoint.sol";
+import {UserOperation} from "account-abstraction/contracts/interfaces/IEntryPoint.sol";
+
+import {ECDSA} from "openzeppelin-contracts/contracts/utils/cryptography/ECDSA.sol";
+
+interface ISafe {
+    function enableModule(address module) external;
+
+    function execTransactionFromModule(
+        address to,
+        uint256 value,
+        bytes memory data,
+        uint8 operation
+    ) external returns (bool success);
+}
+
+contract SafeWebAuthnPlugin is BaseAccount {
+    using ECDSA for bytes32;
+
+    address public immutable myAddress;
+    address private immutable _owner;
+    address private immutable _entryPoint;
+
+    address internal constant _SENTINEL_MODULES = address(0x1);
+
+    constructor(address entryPointAddress, address ownerAddress) {
+        myAddress = address(this);
+        _owner = ownerAddress;
+        _entryPoint = entryPointAddress;
+    }
+
+    function validateUserOp(
+        UserOperation calldata userOp,
+        bytes32 userOpHash,
+        uint256 missingAccountFunds
+    ) external override returns (uint256 validationData) {
+        address payable safeAddress = payable(userOp.sender);
+        ISafe senderSafe = ISafe(safeAddress);
+
+        if (missingAccountFunds != 0) {
+            senderSafe.execTransactionFromModule(
+                _entryPoint,
+                missingAccountFunds,
+                "",
+                0
+            );
+        }
+
+        validationData = _validateSignature(userOp, userOpHash);
+    }
+
+    function execTransaction(
+        address to,
+        uint256 value,
+        bytes calldata data
+    ) external payable {
+        address payable safeAddress = payable(msg.sender);
+        ISafe safe = ISafe(safeAddress);
+        require(
+            safe.execTransactionFromModule(to, value, data, 0),
+            "tx failed"
+        );
+    }
+
+    function enableMyself() public {
+        ISafe(address(this)).enableModule(myAddress);
+    }
+
+    function entryPoint() public view override returns (IEntryPoint) {
+        return IEntryPoint(_entryPoint);
+    }
+
+    function owner() public view returns (address) {
+        return _owner;
+    }
+
+    function _validateSignature(
+        UserOperation calldata userOp,
+        bytes32 userOpHash
+    ) internal override returns (uint256 validationData) {
+        bytes32 hash = userOpHash.toEthSignedMessageHash();
+        if (_owner != hash.recover(userOp.signature))
```

Comment:
Is the idea that this plugin is currently validating sigs with ECDSA signatures but you will eventually turn this into webauthn sigs?  Do you think there's any benefit to keeping this as an EOA pluggin and then just creating the webAuthn one separately?

Reply:
> Is the idea that this plugin is currently validating sigs with ECDSA signatures but you will eventually turn this into webauthn sigs?

Yeah the original plan was to swap this out for webauthn sigs. 

> Do you think there's any benefit to keeping this as an EOA pluggin and then just creating the webAuthn one separately?

I don't have a strong opinion. For my work rn, it's probably doesn't add much benefit to have the ECSDA sig plugin too.  I know James is looking into composable verification logic. In that case, it might be useful for testing and building that. I can leave it separate for now and then it's there if we need it :)

Reply:
>  another thought is we could rename this project in the mono repo to be safe-plugins rather than webauthn

Actually I don't think this has been posted in the chat yet but something James and I discussed last week was monorepo structure. I think the rough plan could be:
- Use the webauth directory now as the integrations directory, as it already has foundry and hardhat setup.
    - The integrations directory would then have directories for Safe, Kernel and any other project we integrated with.
    - We'd build our higher level modules here. So both our current work would live in this 'integrations' directory
- Then we'd have a plain foundry project for the low level cryptographic logic e.g. bls and webauthn - that could be used by the high level kernel and safe modules

**TLDR:** similar to what you outlined but have the kernel and safe plugins in one directory (hardhat & foundry), and then have the low level cryptography stuff in a single dedicated directory (just foundry). What do you think to that idea?

I think James is outlining this now under his standup note titled "proposed github structure".

Reply:
I guess for the ecsda plugin, under the potential new structure, that logic would live in the low level crypto dir, and then could be used by the kernel and safe plugins in the integrations dir